src/dune_rules/scope.ml

```diff
@@ -52,59 +52,64 @@ module DB = struct
   end
 
   let create_db_from_stanzas ~instrument_with ~parent ~lib_config stanzas =
-    let (map : Found_or_redirect.t Lib_name.Map.t) =
-      List.map stanzas ~f:(fun stanza ->
-        match (stanza : Library_related_stanza.t) with
-        | Library_redirect s ->
-          let old_public_name = Lib_name.of_local s.old_name in
-          Found_or_redirect.redirect old_public_name s.new_public_name
-        | Deprecated_library_name s ->
-          let old_public_name = Deprecated_library_name.old_public_name s in
-          Found_or_redirect.redirect old_public_name s.new_public_name
-        | Library (dir, (conf : Library.t)) ->
-          let info = Library.to_lib_info conf ~dir ~lib_config |> Lib_info.of_local in
-          Library.best_name conf, Found_or_redirect.found info)
-      |> Lib_name.Map.of_list_reducei ~f:(fun name (v1 : Found_or_redirect.t) v2 ->
-        let res =
-          match v1, v2 with
-          | Found info1, Found info2 -> Error (Lib_info.loc info1, Lib_info.loc info2)
-          | Found info, Redirect (loc, _) | Redirect (loc, _), Found info ->
-            Error (loc, Lib_info.loc info)
-          | Redirect (loc1, lib1), Redirect (loc2, lib2) ->
-            if Lib_name.equal lib1 lib2 then Ok v1 else Error (loc1, loc2)
-        in
-        match res with
-        | Ok x -> x
-        | Error (loc1, loc2) ->
-          let main_message =
-            Pp.textf "Library %s is defined twice:" (Lib_name.to_string name)
-          in
-          let annots =
-            let main = User_message.make ~loc:loc2 [ main_message ] in
-            let related =
-              [ User_message.make ~loc:loc1 [ Pp.text "Already defined here" ] ]
+    let open Memo.O in
+    let map =
+      Memo.lazy_ (fun () ->
```

Comment:
Is this performance optimization related in some way to the main purpose of the PR?